src/Features/LanguageServer/Protocol/RequestDispatcher.cs

```diff
@@ -8,44 +8,49 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using CommonLanguageServerProtocol.Framework;
 using Microsoft.CodeAnalysis.LanguageServer.Handler;
 using Roslyn.Utilities;
 using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.CodeAnalysis.LanguageServer
 {
-    internal readonly record struct RequestHandlerMetadata(string MethodName, Type RequestType, Type ResponseType);
-
     /// <summary>
     /// Aggregates handlers for the specified languages and dispatches LSP requests
     /// to the appropriate handler for the request.
     /// </summary>
-    internal class RequestDispatcher : ILspService
+    public class RequestDispatcher<RequestContextType> : IRequestDispatcher<RequestContextType>, ILspService where RequestContextType : struct
```

Comment:
Why not in the CLaSP layer?

Reply:
Same as above, gonna send it there eventually.

Reply:
Not sure if GitHub was hiding it from you or it only only showed up in a later commit, but [here's a direct link](https://github.com/dotnet/roslyn/pull/62924/files#diff-41fc8152852a6a741919a988fc6db226f2e54cf4854fb4939b92dd9ebf72f458).

@dibarbet is the crux of your questioning that IRequestDispatcher doesn't do THAT much of interest and could be combined with `RequestExecutionQueue`? I see `IRequestDispatcher`'s job as "Decide where to send this and get it ready", whereas `IRequestExecutionQueue` is more of "Manage the lifetime of requests and schedule them properly" if that pushes your thinking one way or another.

Reply:
Ah I see it now, not sure why I couldn't find it before.  I guess it relates to my overall question of what our API surface should actually be, and how much is extensible which we can discuss separately